contracts/ERC1238/extensions/ERC1238Collection.sol

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "../ERC1238.sol";
+import "./IERC1238Collection.sol";
+
+/**
+ * @dev See {IERC1238Collection}.
+ * This contract keeps track of the balances of addresses for each baseId.
+ * Values are packed the following way in the id:
+ * [baseId (48 bits)][owner (160 bits)][counter (48 bits)]
```

Comment:
Yeh I think it'll be really important to add a lot of comments or explanations on how minting should work, how IDs should be selected to be used properly, how they should be tracked, and how each component of the bit-packing contributes to the delineation of fungible tokens and their relationship with the non-fungible.

Reply:
Sure!
- baseId is needed obviously as reference to the collection
- token owner creates a convenient namespacing and we know that it won't change bc the tokens are non-tansferable
- counter is there because let's say you have a game and you receive tokens when you complete level 1 (so baseId=1), you might redo this level to get a better score (more tokens) so baseId+owner alone isn't enough, a counter helps to make each of your achievements unique while still belonging to the same collection.

Reply:
Thanks for the extra info.

Just continuing on that. How does this interact with the `_balances` mapping in the core implementation and what is their relationship i.e. what does each mapping handle and where are they exclusive from each other?

Reply:
It just seems that, even without this extension, the tokens were already semi-fungible:

```solidity
mapping(uint256 => (address => uint256)) _balances;


_balances[TokenA][user0] = 1; // for NFTs
_balances[TokenB][user0] = 5; // for FTs
```

Again we reach the point where I still think there's a collision in featureset because in the NFT case, TokenId is a token ID but in the FT TokenId acts as a collection ID.

If we are including this kind of flexibility for adding fungible/non-fungibles in parallel, we should support it properly using the struct suggestion I made before.

If not, then we should be purist here and just focus on the functionality of Non-Transferability and not attempt to also introduce new token mapping/structure designs.

To push this argument through, I think we should be more open to the idea of what "Badges" can be and all the different forms it could take which are just a single use-case for the NTT standard; otherwise we are just building a Badge standard, not a general NTT standard. Let's be pure and clear in our contextual approach and ensure we aren't being opinionated or one-sided about features/implementation.

Sorry for rant. :D

Reply:
I feel like there's just growing complexity that is obfuscated behind growing assumptions about specific use-cases as opposed to it being advised by a simple approach that is abstract and general enough for it to be given meaning by implementers/deployers. And if we really want to build-in some functionality we should be very clear about why and how, and build it in a bit more thoroughly than just through bit-packing which I find it to be quite unfriendly.

Reply:
> By semi-fungible I mean that an NFT can have a unique id and at the same time be identified as part of a collection.

Right but this is your definition. I illustrate this because people, similar to myself, could also interpret "semi-fungibility" different from you, and also see the possibility of using the base contract to encode some fungible collections next to non-fungible tokens and call it semi-fungible.

To alleviate this, we have to be very very conceptually clean and make sure there's not any conceptual ambiguity which means not mixing them. Either pick a token type and have the base be `mapping(address => uint256) _balances` or `mapping(uint256 => address) _ownership` OR handle both to the same degree `mapping(uint256 => TokenCollection) collections` and each collection records ownership/balance of that token type.

The arbitrary mix of functionality and design choice is a source of confusion for me and I reckon will be a source of confusion for some others.

> Yes, for FTs, token ids are still token ids but it's true they can serve as collection ID but not for NFTs, hence why this extension is needed.

This asymmetry is exactly why we should pick a side. It doesn't really make sense why we would support one type of token more than another as a design goal which is why it seems more like it's a technical drawback of the chosen approach. So let's just choose a better approach that doesn't render us an imperfect solution without a real reason.

> I think it's about providing optional features, the same way OZ lets you add features to your ERC721 contract.

Right but it still has complete features without the extensions as opposed to partial ones here.

> I don't follow because to me badges and NTTs are just 2 different names for the same thing but I'm curious to know what distinction you see.

Yeh, I completely saw this which is why I mentioned it. Let's go back to basics. NTT stands for Non-Transferable Token. As its most basic it could simply be an ERC20 without `transfer` functions or an ERC721 without `transfer` functions. At its most complex, it could be something completely different and verbose and "Badges" sit somewhere in between. Here we are trying to fit token-type agnosticism with non-transferability into a standard that seems to be more driven by a single example conceptual construction of "Badges", whereas I think we should be thinking about each additional feature to the standard in isolation, discuss its merit & intent and the best approach, as opposed to framing it from a specific use-case and using it to explain technical drawbacks of the approach. In this case its obviously clear why an extension would be relevant but only from that perspective.

I definitely see the case for "Badge" where you want non-fungible token types to have fungible balances, but that's beyond _just_ a non-transferable token, and so I would even go as far as to just call it a Badge Token Standard (BTS, kpop lovers would rave for this :D). As someone arriving expecting a simple standard for a token that can't be transferred and seeing all the specific use-case complexity for this Badge thingy that I didn't exactly want it for, it'd be terribly disappointed.

It's not that I have a problem with the Badge use-case/implementation itself, but I feel that there's conflation between NTT and Badges as pure concepts and an intertwining of use-cases; the case for semi-fungibles in the Badge scenario doesn't seem nearly as strong. The problem I see is potential the desire to combine a multi-token standard into a non-transferable token standard and maybe what we actually want is a token-type agnostic standard where someone can plug their favourite token type (20, 223, 721, 777, 1155 etc.) with the standard and introduce non-transferability logic to it.

> I think on the contrary that bit-packing is a neat/elegant solution, what makes it unfriendly in your opinion?

It's neat, but not elegant, because its done under the assumption that the other design choices are fixed, which they aren't, and so instead of creating this complexity with the stance that its unavoidable due to the base structures, just change the base structures.

Again just going back to a previous point, there's a mix between decisive design choice and inevitable design choice and both arguments are being made at different layers. What makes it confusing to see _why_ things are done a certain way will not make it obvious what is gained by doing it that way.

Concretely, its complex for no specific design reason other than to fit into what was made available by the core layer. If such a hack needs to be performed in order to create such an extension, it leads me to see (from a design perspective) that the core is simply not easily compatible with an extension of this sort in its current form. I could create similar designs for two completely incompatible modules, and find a very specific, fragile way to make them compatible, but if I were to redesign them, I would design it with compatibility in mind. So I think this is what we should do.

---

Sorry again for rant!!!! I hope you're enjoying the engagement, and apologies if these should have been voiced earlier in the development process!!!

Reply:
As a record of our conversation
- Benefit of modelling after ERC1155 = an already familiar design and inherits the design choices
- Let's bring this conversation to the broader community
- We could potentially have different flavors of ERC1238 modelled after ERC20, 721, 1155